Use the estimated SNR per-spw to spwmap low SNR spws, regardless of whether they succeeded

psheehan opened the pull request:
There were at least a few cases found in operations where selfcal encountered datasets with very narrow spectral windows (particularly after cont.dat was applied) but that still succeed at getting selfcal solutions in spite of this, but often the selfcal solutions can cause problems with the data (e.g. in one case we see the source actually move). Notably, it seems that there may be some issues with the SNRs for the gaincal solutions, because in some cases the SNRs of the solutions are comparable to SPWs with much larger bandwidths (including after cont.dat application).

This PR is to have selfcal use the estimated SNR per-spw to decide whether or not it is safe to solve for the gains using that SPW, or whether to SPW map it instead. This should avoid the above problem, as the code will no longer attempt to self-calibrate low SNR SPWs using themselves, and so any oddities in the gaincal SNRs should be mitigated.

jjtobin commented on selfcal_helpers.py:
```diff
@@ -1187,19 +1187,16 @@ def get_SNR_self_individual(vislist,selfcal_library,n_ant,solints,integration_ti
                SNR_self_EB[i]=SNR/((n_ant)**0.5*(selfcal_library['Total_TOS']/selfcal_library[selfcal_library['vislist'][i]]['TOS'])**0.5)
                SNR_self_EB_spw[selfcal_library['vislist'][i]]={}
                for spw in selfcal_library[selfcal_library['vislist'][i]]['spwsarray']:
-                  if spw in SNR_self_EB_spw[selfcal_library['vislist'][i]].keys():
```
I don't think this if statement should be removed. This will cause issues, particularly VLA data, where you might have different numbers of spws for executions of the same SB because some baseline boards were unavailable and there were less than 64 spws possible.

psheehan replied:
Ah, good catch - I'm not actually sure why I took this out in the first place, but will put it back in and run some tests.

psheehan replied:
Ok, I took a closer look at this, and the reason I removed the if statement is because a few lines above SNR_self_EB_spw is defined to be {}. So the if statement can never evaluate to true because the line to populate spw in SNR_self_EB_spw is inside the if statement.

I *think* that the for loop above the if statement, though, should handle spw lists that vary between EBs? Because selfcal_library[vis]['spwsarray'] should only contain the spws in vis.

That said... now looking at the lines below, there might be issues when spws change between EBs. Will mark those lines with a separate comment in a moment in case we want to discuss this particular line further here.

psheehan commented:
Ok, @jjtobin, I've fixed up the issue I noted yesterday. Now the solint_snr_per_spw dictionary uses the virtual spw ids for storing the estimated SNR per spw, and this is all working on the original dataset that motivated this ticket as well as the mixed_spw_EBs ticket, where the spws do change per-EB, to make sure that those updates worked properly.

As an aside mostly so you are aware - as a part of this, I added a 'reverse_spw_map' to the selfcal_library to go from true spw -> virtual spw more easily.

I compared mixed_spw_EBs on this branch vs. on your main and am seeing some slight differences in the outcome, but I think they are minor and not bad. For both sources, as a number of the spws are narrow, they now go to combinespw for inf_EB rather than spwmap. This leads to K04166 slightly failing inf_EB whereas it ever so slightly passed before. But since the improvement was tiny, I don't think this is a big deal. For K04169, interestingly the per-solint improvement actually gets better when looking at DR_post/DR_pre, but the reported final DR goes down ever so slightly. I suspect slight differences in masks leading to slightly different RMS values, and again this is small, so I don't think this is a big deal.

What I did find interesting is that the estimated SNRs that we calculate:

inf_EB: 5.86
inf_EB: spw: 0: 2.36
inf_EB: spw: 1: 0.52
inf_EB: spw: 2: 0.53
inf_EB: spw: 3: 2.28
inf_EB: spw: 4: 3.80

Are quite a bit off from what gaincal reports for the actual solutions (values from ~5-40, and I was a bit surprised by the discrepancy. I know ours is just an estimate... but thats still fairly different. If I had to guess, the source might be resolved in the visibility plane and the solutions are getting a boost from the high SNR short baseliness... but I haven't investigated carefully. @jjtobin, you may done more testing of these functions than I have, so is this something you've encountered/thought about? Based on what I saw from the original dataset that spawned this ticket, I'm probably more suspicious of the gaincal SNRs (as for that dataset, the *very* narrow spws had similar SNRs to much wider ones). The runs are here, in case you want to take a look:

/lustre/cv/projects/SRDP/selfcal-prototyping/Patrick/mixed_spw_EBs_john_main_08_30_2024T10_33_40/
/lustre/cv/projects/SRDP/selfcal-prototyping/Patrick/mixed_spw_EBs_spwmap_based_on_per_spw_SNR_08_30_2024T08_16_15/

In any case, that may not be something that needs to be fixed for this ticket, but thought I'd point it out. Let me know if you have any other questions/concerns/comments on these fixes!

psheehan commented:
Ok, I had to make one further update. Ironically, when I moved to the virtual spws, I needed to put the catch for whether an spw was in an EB back in. But now it is working on the VLA dataset that you pointed me to, along with the two original datasets that I tested. The VLA dataset run result is identical to a run on your main. The runs are here:

/lustre/cv/projects/SRDP/selfcal-prototyping/Patrick/IRAS04169_john_main_09_03_2024T12_48_24
/lustre/cv/projects/SRDP/selfcal-prototyping/Patrick/IRAS04169_spwmap_based_on_per_spw_SNR_09_03_2024T12_54_03

in case you want to take a look. The original two test datasets are identical to their results before this last update.

So I think this should be good to go unless you have any further thoughts!
